### PR TITLE
:sparkles: add key wrapping/unwrapping

### PR DESCRIPTION
**Pull request by heavenboy8**
Adding the feature allowing to wrap and unwrap keys

### CODE REVIEW

File: src/wrapping.rs [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+const DEFAULT_IV: u64 = 0xa6a6a6a6a6a6a6a6;
+
+/// Build the iv according to the RFC 5649
+///
+/// :data_size: is the size of the key to wrap
+fn build_iv(data_size: usize) -> u64 {
+    let l = 0xA65959A6u32;
+    let r = u32::to_le(data_size as u32);
+
+    (l as u64) << 32 | (r as u64)
```
Comment (Manuthor) [inline src/wrapping.rs:17]:
```suggestion
    u64::from(l) << 32 | u64::from(r)
```

File: src/wrapping.rs [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+    let mut R: Vec<u64> = vec![];
+
+    for chunck in plain.chunks(8) {
+        R.push(u64::from_be_bytes(chunck.try_into()?));
+    }
+
+    for j in 0..6 {
+        for (i, r) in R.iter_mut().enumerate().take(n) {
+            // B = AES(K, A | R[i])
+            let I: u128 = (A as u128) << 64 | (*r as u128);
```
Comment (Manuthor) [inline src/wrapping.rs:195]:
```suggestion
            let I: u128 = u128::from(A) << 64 | u128::from(*r);
```

File: src/wrapping.rs [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+    }
+
+    let mut A = R[0];
+
+    for j in (0..6).rev() {
+        for (i, r) in R[1..].iter_mut().rev().enumerate().take(n) {
+            let t = ((n * j) + (n - i)) as u64;
+
+            // B = AES-1(K, (A ^ t) | R[i]) where t = n*j+i
+            let I: u128 = ((A ^ t) as u128) << 64 | (*r as u128);
```
Comment (Manuthor) [inline src/wrapping.rs:281]:
```suggestion
            let I: u128 = u128::from(A ^ t) << 64 | u128::from(*r);
```

File: src/lib.rs [hunk 4]
```diff
@@ -7,6 +7,7 @@ pub mod entropy;
 pub mod hybrid_crypto;
 pub mod kdf;
 pub mod symmetric_crypto;
+pub mod wrapping;
```
Comment (Manuthor) [inline src/lib.rs:10]: maybe rename the module to `key_wrapping`?

File: src/wrapping.rs [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+    if (iv >> 32) as u32 != 0xA65959A6u32 {
+        return false;
+    }
+
+    let real_data_size = (iv as u32) as usize;
+    if real_data_size > data_size || real_data_size as usize <= (data_size - 8) {
+        return false;
+    }
+
+    let paddig_size = data_size - real_data_size;
```
Comment (Manuthor) [inline src/wrapping.rs:32]:
```suggestion
    let padding_size = data_size - real_data_size;
```

**Review by Manuthor — COMMENTED** [review]
well done and quickly!

File: src/wrapping.rs [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+    let l = 0xA65959A6u32;
+    let r = u32::to_le(data_size as u32);
+
+    (l as u64) << 32 | (r as u64)
+}
+
+/// Check if an iv value is appropriate according to the RFC 5649
+fn check_iv(iv: u64, data: &[u8]) -> bool {
+    let data_size: usize = data.len();
+    if (iv >> 32) as u32 != 0xA65959A6u32 {
```
Comment (ThibsG) [inline src/wrapping.rs:23]: Maybe we can define a const for the value `0xA65959A6u32`

File: src/wrapping.rs [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+        return false;
+    }
+
+    let real_data_size = (iv as u32) as usize;
+    if real_data_size > data_size || real_data_size as usize <= (data_size - 8) {
```
Comment (ThibsG) [inline src/wrapping.rs:28]: `real_data_size` is already `usize` no?
```suggestion
    if real_data_size > data_size || real_data_size <= (data_size - 8) {
```

File: src/wrapping.rs [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+///
+/// :data_size: is the size of the key to wrap
+fn build_iv(data_size: usize) -> u64 {
+    let l = 0xA65959A6u32;
+    let r = u32::to_le(data_size as u32);
+
+    (l as u64) << 32 | (r as u64)
+}
+
+/// Check if an iv value is appropriate according to the RFC 5649
```
Comment (ThibsG) [inline src/wrapping.rs:20]:
```suggestion
/// Check if `iv` value is appropriate according to the RFC 5649
```

File: src/wrapping.rs [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+    if real_data_size > data_size || real_data_size as usize <= (data_size - 8) {
+        return false;
+    }
+
+    let paddig_size = data_size - real_data_size;
+    if data[real_data_size..] != vec![0u8; paddig_size] {
+        return false;
+    }
+
+    true
```
Comment (ThibsG) [inline src/wrapping.rs:37]:
```suggestion
    // check padding is zeroed
    data[real_data_size..].iter().all(|&x| x == 0)
```

File: src/wrapping.rs [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+}
+
+/// Wrap a plain text of variable length
+///
+/// Following RFC 5649
+#[allow(non_camel_case_types)]
+#[allow(non_snake_case)]
```
Comment (ThibsG) [inline src/wrapping.rs:44]: Can we try to get rid of those (for the whole file)?

I understand this is done to stick with RFC, but reading variables with one letter makes the reading really difficult IMO

File: src/wrapping.rs [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+
+/// Wrap a plain text of a 64-bits modulo size
+///
+/// Following RFC 3394  
+#[allow(non_camel_case_types)]
+#[allow(non_snake_case)]
+fn _wrap_64(plain: &[u8], kek: &[u8], iv: Option<u64>) -> Result<Vec<u8>, CryptoBaseError> {
+    let n = plain.len();
+
+    if plain.len() % 8 != 0 {
```
Comment (ThibsG) [inline src/wrapping.rs:176]:
```suggestion
    if n % 8 != 0 {
```

File: src/wrapping.rs [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+        return Err(CryptoBaseError::InvalidSize(
+            "The plaintext size should be a multiple of 8".to_string(),
+        ));
+    }
+
+    // Number of 64-bit blocks
+    let n = n / 8;
+
+    let mut A = iv.unwrap_or(DEFAULT_IV);
+    let mut R: Vec<u64> = vec![];
```
Comment (ThibsG) [inline src/wrapping.rs:186]:
```suggestion
    let mut R = Vec::with_capacity(n);
```

File: src/wrapping.rs [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+    if n % 8 != 0 || n < 16 {
+        return Err(CryptoBaseError::InvalidSize(
+            "The ciphertext size should be >= 16 and a multiple of 8".to_string(),
+        ));
+    }
+
+    // Number of 64-bit blocks
+    let n = n / 8 - 1;
+
+    let mut R: Vec<u64> = vec![];
```
Comment (ThibsG) [inline src/wrapping.rs:269]:
```suggestion
    let mut R = Vec::with_capacity(n);
```

Comment (heavenboy8) [reply src/wrapping.rs:269]: n+1 in that case

File: src/wrapping.rs (hunk 10 above)
Comment (heavenboy8) [reply src/wrapping.rs:44]: As discuss, we will stick to the RFC variables names

**Pull request merged**